# Import Libraries

In [ ]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder, OrdinalEncoder

jelaskan di sini fungsi setiap library

# Load Data

In [2]:
train = pd.read_csv("../data/train.csv")
test = pd.read_csv("../data/test.csv")

train_df = train.copy()
test_df = test.copy()


# Dataset Overview

## Head

In [3]:
train_df.head()

id      Name  Gender   Age           City Working Professional or Student  \
0   0  Aaradhya  Female  49.0       Ludhiana            Working Professional   
1   1     Vivan    Male  26.0       Varanasi            Working Professional   
2   2    Yuvraj    Male  33.0  Visakhapatnam                         Student   
3   3    Yuvraj    Male  22.0         Mumbai            Working Professional   
4   4      Rhea  Female  30.0         Kanpur            Working Professional   

         Profession  Academic Pressure  Work Pressure  CGPA  \
0              Chef                NaN            5.0   NaN   
1           Teacher                NaN            4.0   NaN   
2               NaN                5.0            NaN  8.97   
3           Teacher                NaN            5.0   NaN   
4  Business Analyst                NaN            1.0   NaN   

   Study Satisfaction  Job Satisfaction     Sleep Duration Dietary Habits  \
0                 NaN               2.0  More than 8 hours        Healthy   
1                 NaN               3.0  Less than 5 hours      Unhealthy   
2                 2.0               NaN          5-6 hours        Healthy   
3                 NaN               1.0  Less than 5 hours       Moderate   
4                 NaN               1.0          5-6 hours      Unhealthy   

    Degree Have you ever had suicidal thoughts ?  Work/Study Hours  \
0      BHM                                    No               1.0   
1      LLB                                   Yes               7.0   
2  B.Pharm                                   Yes               3.0   
3      BBA                                   Yes              10.0   
4      BBA                                   Yes               9.0   

   Financial Stress Family History of Mental Illness  Depression  
0               2.0                               No           0  
1               3.0                               No           1  
2               1.0                               No           1  
3               1.0                              Yes           1  
4               4.0                              Yes           0

In [4]:
test_df.head()

id     Name  Gender   Age           City  \
0  140700   Shivam    Male  53.0  Visakhapatnam   
1  140701    Sanya  Female  58.0        Kolkata   
2  140702     Yash    Male  53.0         Jaipur   
3  140703   Nalini  Female  23.0         Rajkot   
4  140704  Shaurya    Male  47.0         Kalyan   

  Working Professional or Student              Profession  Academic Pressure  \
0            Working Professional                   Judge                NaN   
1            Working Professional  Educational Consultant                NaN   
2            Working Professional                 Teacher                NaN   
3                         Student                     NaN                5.0   
4            Working Professional                 Teacher                NaN   

   Work Pressure  CGPA  Study Satisfaction  Job Satisfaction  \
0            2.0   NaN                 NaN               5.0   
1            2.0   NaN                 NaN               4.0   
2            4.0   NaN                 NaN               1.0   
3            NaN  6.84                 1.0               NaN   
4            5.0   NaN                 NaN               5.0   

      Sleep Duration Dietary Habits  Degree  \
0  Less than 5 hours       Moderate     LLB   
1  Less than 5 hours       Moderate    B.Ed   
2          7-8 hours       Moderate  B.Arch   
3  More than 8 hours       Moderate     BSc   
4          7-8 hours       Moderate     BCA   

  Have you ever had suicidal thoughts ?  Work/Study Hours  Financial Stress  \
0                                    No               9.0               3.0   
1                                    No               6.0               4.0   
2                                   Yes              12.0               4.0   
3                                   Yes              10.0               4.0   
4                                   Yes               3.0               4.0   

  Family History of Mental Illness  
0                              Yes  
1                               No  
2                               No  
3                               No  
4                               No

In [5]:
train_df.tail()

id     Name  Gender   Age       City  \
140695  140695    Vidya  Female  18.0  Ahmedabad   
140696  140696     Lata  Female  41.0  Hyderabad   
140697  140697  Aanchal  Female  24.0    Kolkata   
140698  140698   Prachi  Female  49.0   Srinagar   
140699  140699      Sai    Male  27.0      Patna   

       Working Professional or Student         Profession  Academic Pressure  \
140695            Working Professional                NaN                NaN   
140696            Working Professional     Content Writer                NaN   
140697            Working Professional  Marketing Manager                NaN   
140698            Working Professional            Plumber                NaN   
140699                         Student                NaN                4.0   

        Work Pressure  CGPA  Study Satisfaction  Job Satisfaction  \
140695            5.0   NaN                 NaN               4.0   
140696            5.0   NaN                 NaN               4.0   
140697            3.0   NaN                 NaN               1.0   
140698            5.0   NaN                 NaN               2.0   
140699            NaN  9.24                 1.0               NaN   

           Sleep Duration Dietary Habits    Degree  \
140695          5-6 hours      Unhealthy  Class 12   
140696          7-8 hours       Moderate    B.Tech   
140697  More than 8 hours       Moderate     B.Com   
140698          5-6 hours       Moderate        ME   
140699  Less than 5 hours        Healthy       BCA   

       Have you ever had suicidal thoughts ?  Work/Study Hours  \
140695                                    No               2.0   
140696                                   Yes               6.0   
140697                                    No               4.0   
140698                                   Yes              10.0   
140699                                   Yes               2.0   

        Financial Stress Family History of Mental Illness  Depression  
140695               4.0                              Yes           1  
140696               5.0                              Yes           0  
140697               4.0                               No           0  
140698               1.0                               No           0  
140699               3.0                              Yes           1

In [6]:
test_df.tail()

id    Name  Gender   Age           City  \
93795  234495    Zoya  Female  49.0         Jaipur   
93796  234496   Shlok    Male  29.0      Ahmedabad   
93797  234497   Rishi    Male  24.0  Visakhapatnam   
93798  234498  Eshita  Female  23.0         Kalyan   
93799  234499   Gauri  Female  43.0       Varanasi   

      Working Professional or Student              Profession  \
93795            Working Professional                   Pilot   
93796            Working Professional                   Pilot   
93797                         Student                     NaN   
93798            Working Professional       Marketing Manager   
93799            Working Professional  Educational Consultant   

       Academic Pressure  Work Pressure  CGPA  Study Satisfaction  \
93795                NaN            3.0   NaN                 NaN   
93796                NaN            5.0   NaN                 NaN   
93797                1.0            NaN  7.51                 4.0   
93798                NaN            4.0   NaN                 NaN   
93799                NaN            5.0   NaN                 NaN   

       Job Satisfaction     Sleep Duration Dietary Habits  Degree  \
93795               5.0  Less than 5 hours       Moderate     BSc   
93796               1.0          7-8 hours       Moderate      BE   
93797               NaN          7-8 hours       Moderate  B.Tech   
93798               2.0          5-6 hours        Healthy      BA   
93799               2.0  More than 8 hours        Healthy    B.Ed   

      Have you ever had suicidal thoughts ?  Work/Study Hours  \
93795                                   Yes               2.0   
93796                                   Yes              11.0   
93797                                    No               7.0   
93798                                   Yes               7.0   
93799                                    No              11.0   

       Financial Stress Family History of Mental Illness  
93795               2.0                              Yes  
93796               3.0                              Yes  
93797               1.0                               No  
93798               5.0                              Yes  
93799               2.0                               No

## Info

In [7]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 140700 entries, 0 to 140699
Data columns (total 20 columns):
 #   Column                                 Non-Null Count   Dtype  
---  ------                                 --------------   -----  
 0   id                                     140700 non-null  int64  
 1   Name                                   140700 non-null  object 
 2   Gender                                 140700 non-null  object 
 3   Age                                    140700 non-null  float64
 4   City                                   140700 non-null  object 
 5   Working Professional or Student        140700 non-null  object 
 6   Profession                             104070 non-null  object 
 7   Academic Pressure                      27897 non-null   float64
 8   Work Pressure                          112782 non-null  float64
 9   CGPA                                   27898 non-null   float64
 10  Study Satisfaction                     27897 non-null   

## Describe

In [8]:
train_df.describe()

id            Age  Academic Pressure  Work Pressure  \
count  140700.000000  140700.000000       27897.000000  112782.000000   
mean    70349.500000      40.388621           3.142273       2.998998   
std     40616.735775      12.384099           1.380457       1.405771   
min         0.000000      18.000000           1.000000       1.000000   
25%     35174.750000      29.000000           2.000000       2.000000   
50%     70349.500000      42.000000           3.000000       3.000000   
75%    105524.250000      51.000000           4.000000       4.000000   
max    140699.000000      60.000000           5.000000       5.000000   

               CGPA  Study Satisfaction  Job Satisfaction  Work/Study Hours  \
count  27898.000000        27897.000000     112790.000000     140700.000000   
mean       7.658636            2.944940          2.974404          6.252679   
std        1.464466            1.360197          1.416078          3.853615   
min        5.030000            1.000000          1.000000          0.000000   
25%        6.290000            2.000000          2.000000          3.000000   
50%        7.770000            3.000000          3.000000          6.000000   
75%        8.920000            4.000000          4.000000         10.000000   
max       10.000000            5.000000          5.000000         12.000000   

       Financial Stress     Depression  
count     140696.000000  140700.000000  
mean           2.988983       0.181713  
std            1.413633       0.385609  
min            1.000000       0.000000  
25%            2.000000       0.000000  
50%            3.000000       0.000000  
75%            4.000000       0.000000  
max            5.000000       1.000000

## Missing Value

In [9]:
def missing_data(df):
    """
    Menghitung jumlah total dan persentase nilai yang hilang (missing values) di setiap kolom pada DataFrame.
    
    Parameters:
    df : pd.DataFrame
        DataFrame yang akan diperiksa missing values-nya.
    
    Returns:
    pd.DataFrame
        DataFrame yang berisi jumlah dan persentase missing values di setiap kolom.
    """
    total = df.isnull().sum().sort_values(ascending=False)
    percent = (df.isnull().sum() / df.isnull().count()).sort_values(ascending=False)
    
    missing_data = pd.concat([total, percent], axis=1, keys=['Total', 'Missing Percent'])
    missing_data['Missing Percent'] = missing_data['Missing Percent'].apply(lambda x: x * 100)
    
    return missing_data

In [10]:
missing_data(train_df)

Total  Missing Percent
Study Satisfaction                     112803        80.172708
Academic Pressure                      112803        80.172708
CGPA                                   112802        80.171997
Profession                              36630        26.034115
Work Pressure                           27918        19.842217
Job Satisfaction                        27910        19.836532
Dietary Habits                              4         0.002843
Financial Stress                            4         0.002843
Degree                                      2         0.001421
Family History of Mental Illness            0         0.000000
Work/Study Hours                            0         0.000000
Have you ever had suicidal thoughts ?       0         0.000000
id                                          0         0.000000
Sleep Duration                              0         0.000000
Name                                        0         0.000000
Working Professional or Student             0         0.000000
City                                        0         0.000000
Age                                         0         0.000000
Gender                                      0         0.000000
Depression                                  0         0.000000

In [11]:
missing_data(test_df)

Total  Missing Percent
CGPA                                   75034        79.993603
Academic Pressure                      75033        79.992537
Study Satisfaction                     75033        79.992537
Profession                             24632        26.260128
Work Pressure                          18778        20.019190
Job Satisfaction                       18774        20.014925
Dietary Habits                             5         0.005330
Degree                                     2         0.002132
Sleep Duration                             0         0.000000
Financial Stress                           0         0.000000
Work/Study Hours                           0         0.000000
Have you ever had suicidal thoughts ?      0         0.000000
id                                         0         0.000000
Name                                       0         0.000000
Working Professional or Student            0         0.000000
City                                       0         0.000000
Age                                        0         0.000000
Gender                                     0         0.000000
Family History of Mental Illness           0         0.000000

## Imbalance Data

In [12]:
count_class = train_df['Depression'].value_counts()
count_class

Depression
0    115133
1     25567
Name: count, dtype: int64

Terdapat imbalance data, di mana data untuk label 0 lebih banyak, dibandingkan data untuk label 1

pemisahan dilakukan untuk analisa data numerikal dan kategorikal

## Data Correlation

In [13]:
corr = train_df.select_dtypes('number').corr()
corr

id       Age  Academic Pressure  Work Pressure  \
id                  1.000000  0.001850           0.005597       0.003086   
Age                 0.001850  1.000000          -0.075956      -0.098924   
Academic Pressure   0.005597 -0.075956           1.000000            NaN   
Work Pressure       0.003086 -0.098924                NaN       1.000000   
CGPA               -0.011869  0.005129          -0.025195            NaN   
Study Satisfaction  0.007824  0.008862          -0.112228            NaN   
Job Satisfaction    0.002774  0.030793          -1.000000      -0.036958   
Work/Study Hours    0.002359 -0.118356           0.095650      -0.002808   
Financial Stress   -0.000340 -0.083589           0.152105       0.024412   
Depression          0.003944 -0.564671           0.475037       0.216634   

                        CGPA  Study Satisfaction  Job Satisfaction  \
id                 -0.011869            0.007824          0.002774   
Age                 0.005129            0.008862          0.030793   
Academic Pressure  -0.025195           -0.112228         -1.000000   
Work Pressure            NaN                 NaN         -0.036958   
CGPA                1.000000           -0.047078          1.000000   
Study Satisfaction -0.047078            1.000000         -1.000000   
Job Satisfaction    1.000000           -1.000000          1.000000   
Work/Study Hours    0.002875           -0.036934         -0.029227   
Financial Stress    0.006398           -0.065001         -0.029160   
Depression          0.021729           -0.168014         -0.168543   

                    Work/Study Hours  Financial Stress  Depression  
id                          0.002359         -0.000340    0.003944  
Age                        -0.118356         -0.083589   -0.564671  
Academic Pressure           0.095650          0.152105    0.475037  
Work Pressure              -0.002808          0.024412    0.216634  
CGPA                        0.002875          0.006398    0.021729  
Study Satisfaction         -0.036934         -0.065001   -0.168014  
Job Satisfaction           -0.029227         -0.029160   -0.168543  
Work/Study Hours            1.000000          0.036134    0.191746  
Financial Stress            0.036134          1.000000    0.227237  
Depression                  0.191746          0.227237    1.000000

### 1. **Age (Usia)**
**Korelasi dengan variabel lain**:  
- **Age** tidak menunjukkan korelasi yang signifikan dengan variabel lain dalam dataset, dengan sebagian besar korelasi mendekati nol.  

**Interpretasi**:  
- Hal ini menunjukkan bahwa **usia** tidak terlalu berpengaruh dalam menentukan seseorang mengalami depresi, stres keuangan, atau faktor-faktor lain seperti tekanan akademik dan kepuasan studi/kerja. Usia mungkin bukan faktor kunci dalam dataset ini.

---

### 2. **Academic Pressure (Tekanan Akademik)**
**Korelasi dengan variabel lain**:  
- **Depression**: Korelasi positif tinggi (0.4750) menunjukkan bahwa semakin tinggi tekanan akademik, semakin besar kemungkinan seseorang mengalami depresi.  
- **Work/Study Hours**: Korelasi positif kecil (0.0957), menunjukkan sedikit hubungan antara jam kerja/studi dengan tekanan akademik.  
- **Financial Stress**: Korelasi positif moderat (0.1521), menunjukkan bahwa individu yang merasakan tekanan akademik juga cenderung mengalami stres keuangan.  

**Interpretasi**:  
- **Academic Pressure** merupakan variabel yang sangat penting dalam dataset ini, karena memiliki korelasi yang tinggi dengan **depresi**. Selain itu, variabel ini juga menunjukkan korelasi positif yang moderat dengan **stres keuangan** dan sedikit dengan **jam kerja/studi**. Oleh karena itu, tekanan akademik mungkin menjadi faktor utama dalam meningkatkan depresi dan stres keuangan.

---

### 3. **Work Pressure (Tekanan Kerja)**
**Korelasi dengan variabel lain**:  
- **Depression**: Korelasi positif moderat (0.2166), menunjukkan bahwa semakin tinggi tekanan kerja, semakin besar kemungkinan seseorang mengalami depresi.  
- **Study Satisfaction**: Korelasi negatif kecil (-0.1122), menunjukkan sedikit hubungan negatif antara tekanan kerja dan kepuasan belajar.  
- **Job Satisfaction**: Korelasi negatif kecil (-0.0369), menunjukkan sedikit hubungan negatif antara tekanan kerja dan kepuasan kerja.  

**Interpretasi**:  
- **Work Pressure** memiliki hubungan yang lebih lemah dibandingkan dengan **academic pressure**. Meskipun ada korelasi positif dengan **depresi**, pengaruhnya lebih kecil. Tekanan kerja tampaknya tidak terlalu berpengaruh terhadap kepuasan studi dan pekerjaan dalam konteks dataset ini.

---

### 4. **CGPA (Indeks Prestasi Kumulatif)**
**Korelasi dengan variabel lain**:  
- **Study Satisfaction**: Korelasi negatif kecil (-0.0471), menunjukkan bahwa ada sedikit hubungan antara CGPA yang lebih tinggi dan kepuasan belajar yang lebih rendah.  
- **Job Satisfaction**: Korelasi positif kecil (0.0217), menunjukkan sedikit hubungan positif antara CGPA dan kepuasan kerja.  

**Interpretasi**:  
- **CGPA** tidak memiliki korelasi yang signifikan dengan banyak variabel lain dalam dataset ini. Hal ini menunjukkan bahwa nilai akademik (IPK) tidak berpengaruh besar terhadap kepuasan studi atau pekerjaan, dan mungkin bukan faktor yang mendalam dalam mempengaruhi stres atau depresi.

---

### 5. **Study Satisfaction (Kepuasan Studi)**
**Korelasi dengan variabel lain**:  
- **Academic Pressure**: Korelasi negatif kecil (-0.1122), menunjukkan bahwa semakin tinggi tekanan akademik, semakin rendah kepuasan studi.  
- **Job Satisfaction**: Korelasi negatif sempurna (-1.0), menunjukkan hubungan yang sangat kuat dan negatif dengan kepuasan pekerjaan.  
- **Depression**: Korelasi negatif kecil (-0.1680), menunjukkan sedikit hubungan antara kepuasan studi yang rendah dan depresi.  

**Interpretasi**:  
- **Study Satisfaction** berhubungan negatif dengan **Academic Pressure** dan **Job Satisfaction**, yang menunjukkan bahwa mereka yang lebih tertekan secara akademik cenderung kurang puas dengan studi mereka. Hubungan negatif dengan **depresi** juga mengindikasikan bahwa kepuasan belajar yang rendah sedikit terkait dengan tingkat depresi yang lebih tinggi.

---

### 6. **Job Satisfaction (Kepuasan Kerja)**
**Korelasi dengan variabel lain**:  
- **Study Satisfaction**: Korelasi negatif sempurna (-1.0), menunjukkan hubungan yang sangat kuat dan negatif dengan kepuasan studi.  
- **Depression**: Korelasi negatif kecil (-0.1685), menunjukkan sedikit hubungan negatif antara kepuasan kerja dan depresi.  

**Interpretasi**:  
- **Job Satisfaction** menunjukkan korelasi negatif yang sangat kuat dengan **Study Satisfaction**, yang menunjukkan bahwa mereka yang lebih puas dengan pekerjaan mereka cenderung tidak puas dengan studi mereka, atau sebaliknya. Kepuasan kerja juga berhubungan sedikit dengan depresi.

---

### 7. **Work/Study Hours (Jam Kerja/Belajar)**
**Korelasi dengan variabel lain**:  
- **Academic Pressure**: Korelasi positif kecil (0.0957), menunjukkan sedikit hubungan antara jam kerja/studi dengan tekanan akademik.  
- **Depression**: Korelasi positif moderat (0.1917), menunjukkan bahwa semakin banyak waktu yang dihabiskan untuk bekerja atau belajar, semakin besar kemungkinan seseorang mengalami depresi.  

**Interpretasi**:  
- **Work/Study Hours** berhubungan sedikit dengan **Academic Pressure**, namun menunjukkan korelasi positif dengan **Depression**. Jam kerja/studi yang lebih tinggi mungkin berkontribusi pada tingkat depresi yang lebih tinggi, meskipun hubungan ini tidak sangat kuat.

---

### 8. **Financial Stress (Stres Keuangan)**
**Korelasi dengan variabel lain**:  
- **Academic Pressure**: Korelasi positif moderat (0.1521), menunjukkan bahwa semakin besar tekanan akademik, semakin besar kemungkinan seseorang mengalami stres keuangan.  
- **Depression**: Korelasi positif kecil (0.2272), menunjukkan sedikit hubungan positif antara stres keuangan dan depresi.  

**Interpretasi**:  
- **Financial Stress** menunjukkan korelasi positif dengan **Academic Pressure** dan **Depression**, yang mengindikasikan bahwa stres keuangan sering kali terkait dengan tekanan akademik dan tingkat depresi yang lebih tinggi.

---

### 9. **Depression (Depresi)**
**Korelasi dengan variabel lain**:  
- **Academic Pressure**: Korelasi positif tinggi (0.4750), menunjukkan bahwa semakin tinggi tekanan akademik, semakin besar kemungkinan seseorang mengalami depresi.  
- **Work Pressure**: Korelasi positif moderat (0.2166), menunjukkan sedikit hubungan antara tekanan kerja dan depresi.  
- **Financial Stress**: Korelasi positif kecil (0.2272), menunjukkan sedikit hubungan antara stres keuangan dan depresi.  

**Interpretasi**:  
- **Depression** memiliki korelasi positif yang moderat dengan **Academic Pressure**, **Work Pressure**, dan **Financial Stress**, menunjukkan bahwa depresi lebih sering dialami oleh mereka yang merasakan tekanan akademik, tekanan kerja, dan stres keuangan. Ini mengindikasikan pentingnya faktor-faktor eksternal dalam mempengaruhi tingkat depresi.

---

### Kesimpulan Umum:
1. **Academic Pressure** adalah faktor paling signifikan yang mempengaruhi **Depression**, serta terkait dengan **Stres Keuangan** dan sedikit dengan **Work/Study Hours**.
2. **Work Pressure** berhubungan sedikit dengan **Depression**, tetapi pengaruhnya lebih kecil dibandingkan dengan tekanan akademik.
3. **Kepuasan Studi** dan **Kepuasan Kerja** sangat saling terkait dan negatif, dengan kepuasan studi yang lebih rendah sering dikaitkan dengan kepuasan kerja yang lebih tinggi.
4. **Depression** sangat dipengaruhi oleh tekanan akademik dan kerja, serta stres keuangan, yang harus menjadi fokus perhatian untuk mengurangi dampak negatif terhadap individu.


# Data Preprocessing

## Handling Missing Value Categorical

In [14]:
cat_treshold = 100

In [15]:
for col in train_df.select_dtypes('object').columns:
    train_df[col] = train_df[col].fillna('missing')
    train_df.loc[train_df[col].value_counts(dropna=False)[train_df[col]].values < cat_treshold, col] = 'noise'

In [16]:
for col in test_df.select_dtypes('object').columns:
    test_df[col] = test_df[col].fillna('missing')
    test_df.loc[test_df[col].value_counts(dropna=False)[test_df[col]].values < cat_treshold, col] = 'noise'

In [17]:
missing_data(train_df)

Total  Missing Percent
Study Satisfaction                     112803        80.172708
Academic Pressure                      112803        80.172708
CGPA                                   112802        80.171997
Work Pressure                           27918        19.842217
Job Satisfaction                        27910        19.836532
Financial Stress                            4         0.002843
Sleep Duration                              0         0.000000
Family History of Mental Illness            0         0.000000
Work/Study Hours                            0         0.000000
Have you ever had suicidal thoughts ?       0         0.000000
Degree                                      0         0.000000
Dietary Habits                              0         0.000000
id                                          0         0.000000
Name                                        0         0.000000
Profession                                  0         0.000000
Working Professional or Student             0         0.000000
City                                        0         0.000000
Age                                         0         0.000000
Gender                                      0         0.000000
Depression                                  0         0.000000

In [18]:
missing_data(test_df)

Total  Missing Percent
CGPA                                   75034        79.993603
Academic Pressure                      75033        79.992537
Study Satisfaction                     75033        79.992537
Work Pressure                          18778        20.019190
Job Satisfaction                       18774        20.014925
Financial Stress                           0         0.000000
Work/Study Hours                           0         0.000000
Have you ever had suicidal thoughts ?      0         0.000000
Degree                                     0         0.000000
Dietary Habits                             0         0.000000
Sleep Duration                             0         0.000000
id                                         0         0.000000
Name                                       0         0.000000
Profession                                 0         0.000000
Working Professional or Student            0         0.000000
City                                       0         0.000000
Age                                        0         0.000000
Gender                                     0         0.000000
Family History of Mental Illness           0         0.000000

In [19]:
num_cols_train = [cols for cols in train_df.select_dtypes('number').columns if cols!= 'id' and cols!='Depression']

In [20]:
for col in num_cols_train:
    train_df[col]=train_df[col].fillna(0)

In [21]:
train_df.isnull().sum()

id                                       0
Name                                     0
Gender                                   0
Age                                      0
City                                     0
Working Professional or Student          0
Profession                               0
Academic Pressure                        0
Work Pressure                            0
CGPA                                     0
Study Satisfaction                       0
Job Satisfaction                         0
Sleep Duration                           0
Dietary Habits                           0
Degree                                   0
Have you ever had suicidal thoughts ?    0
Work/Study Hours                         0
Financial Stress                         0
Family History of Mental Illness         0
Depression                               0
dtype: int64

In [22]:
num_cols_test = [cols for cols in test_df.select_dtypes('number').columns if cols!= 'id']

In [23]:
for col in num_cols_test:
    test_df[col]=test_df[col].fillna(0)

In [24]:
test_df.isnull().sum()

id                                       0
Name                                     0
Gender                                   0
Age                                      0
City                                     0
Working Professional or Student          0
Profession                               0
Academic Pressure                        0
Work Pressure                            0
CGPA                                     0
Study Satisfaction                       0
Job Satisfaction                         0
Sleep Duration                           0
Dietary Habits                           0
Degree                                   0
Have you ever had suicidal thoughts ?    0
Work/Study Hours                         0
Financial Stress                         0
Family History of Mental Illness         0
dtype: int64

## Feature Engineering

In [25]:
train_df['Job/Study Satisfaction']=train_df['Job Satisfaction']+train_df['Study Satisfaction']
test_df['Job/Study Satisfaction']=test_df['Job Satisfaction']+test_df['Study Satisfaction']

In [26]:
train_df['Work/Academic Pressure']=train_df['Work Pressure']+train_df['Academic Pressure']
test_df['Work/Academic Pressure']=test_df['Work Pressure']+test_df['Academic Pressure']

In [27]:
train_df.drop(['Job Satisfaction','Study Satisfaction','Work Pressure', 'Academic Pressure'],inplace=True,axis=1)
test_df.drop(['Job Satisfaction','Study Satisfaction','Work Pressure', 'Academic Pressure'],inplace=True,axis=1)

In [28]:
train_df.columns

Index(['id', 'Name', 'Gender', 'Age', 'City',
       'Working Professional or Student', 'Profession', 'CGPA',
       'Sleep Duration', 'Dietary Habits', 'Degree',
       'Have you ever had suicidal thoughts ?', 'Work/Study Hours',
       'Financial Stress', 'Family History of Mental Illness', 'Depression',
       'Job/Study Satisfaction', 'Work/Academic Pressure'],
      dtype='object')

In [29]:
test_df.columns

Index(['id', 'Name', 'Gender', 'Age', 'City',
       'Working Professional or Student', 'Profession', 'CGPA',
       'Sleep Duration', 'Dietary Habits', 'Degree',
       'Have you ever had suicidal thoughts ?', 'Work/Study Hours',
       'Financial Stress', 'Family History of Mental Illness',
       'Job/Study Satisfaction', 'Work/Academic Pressure'],
      dtype='object')

## Mapping Sleep Duration

In [30]:
train_df['Sleep Duration'].value_counts()

Sleep Duration
Less than 5 hours    38784
7-8 hours            36969
More than 8 hours    32726
5-6 hours            32142
noise                   79
Name: count, dtype: int64

In [31]:
def sleep_duration(row):
    if(row['Sleep Duration']=='Less than 5 hours'):
        return 4
    elif (row['Sleep Duration']=='7-8 hours'):
        return 7.5
    elif (row['Sleep Duration']=='More than 8 hours'):
        return 9
    elif (row['Sleep Duration']=='5-6 hours'):
        return 5.5
    else:
        return 7

In [32]:
train_df['Sleep Duration'] = train_df.apply(sleep_duration, axis=1)

In [33]:
test_df['Sleep Duration'] = test_df.apply(sleep_duration, axis=1)

## Drop Irrelevant Columns

In [34]:
col_to_drop = ['Name', 'City', 'CGPA', 'Degree', 'Profession']
train_df = train_df.drop(columns=col_to_drop)
test_df = test_df.drop(columns=col_to_drop)

In [35]:
print(train_df.columns, '\n')
print(test_df.columns)

Index(['id', 'Gender', 'Age', 'Working Professional or Student',
       'Sleep Duration', 'Dietary Habits',
       'Have you ever had suicidal thoughts ?', 'Work/Study Hours',
       'Financial Stress', 'Family History of Mental Illness', 'Depression',
       'Job/Study Satisfaction', 'Work/Academic Pressure'],
      dtype='object') 

Index(['id', 'Gender', 'Age', 'Working Professional or Student',
       'Sleep Duration', 'Dietary Habits',
       'Have you ever had suicidal thoughts ?', 'Work/Study Hours',
       'Financial Stress', 'Family History of Mental Illness',
       'Job/Study Satisfaction', 'Work/Academic Pressure'],
      dtype='object')


## Drop Irrelevant Values

In [36]:
for col in train_df.columns:
    print(f"Value counts for {col}:")
    print(train_df[col].value_counts())
    print("\n")

Value counts for id:
id
0         1
93803     1
93797     1
93798     1
93799     1
         ..
46893     1
46892     1
46891     1
46890     1
140699    1
Name: count, Length: 140700, dtype: int64


Value counts for Gender:
Gender
Male      77464
Female    63236
Name: count, dtype: int64


Value counts for Age:
Age
56.0    5246
49.0    5099
38.0    4564
53.0    4526
57.0    4395
47.0    4199
46.0    4080
54.0    3928
51.0    3927
18.0    3921
43.0    3856
59.0    3781
48.0    3695
45.0    3535
20.0    3515
50.0    3439
42.0    3393
24.0    3355
28.0    3338
39.0    3262
44.0    3217
41.0    3140
33.0    3140
29.0    3097
40.0    3094
58.0    2932
25.0    2931
23.0    2900
55.0    2846
35.0    2834
21.0    2742
37.0    2724
36.0    2703
19.0    2634
27.0    2613
52.0    2589
34.0    2578
31.0    2560
60.0    2501
26.0    2108
22.0    2066
30.0    1912
32.0    1785
Name: count, dtype: int64


Value counts for Working Professional or Student:
Working Professional or Student
Working Profe

In [37]:
train_df = train_df[train_df['Dietary Habits'] != 'noise']
test_df = test_df[test_df['Dietary Habits'] != 'noise']


In [38]:
# Fill 0 with median for Work/Study Hours (median is eligible for handling outlier)
train_df['Work/Study Hours'] = train_df['Work/Study Hours'].replace(0, train_df['Work/Study Hours'].median())

# Fill 0 with mode for Financial Stress (there are 4 values so we use mode to keep another value dominance)
train_df['Financial Stress'] = train_df['Financial Stress'].replace(0, train_df['Financial Stress'].mode()[0])

# Fill 0 with mode for Job/Study Satisfaction
train_df['Job/Study Satisfaction'] = train_df['Job/Study Satisfaction'].replace(0, train_df['Job/Study Satisfaction'].mode()[0])

# Fill 0 with mode for Work/Academic Pressure
train_df['Work/Academic Pressure'] = train_df['Work/Academic Pressure'].replace(0, train_df['Work/Academic Pressure'].mode()[0])

In [39]:
# Fill 0 with median for Work/Study Hours (median is eligible for handling outlier)
test_df['Work/Study Hours'] = test_df['Work/Study Hours'].replace(0, test_df['Work/Study Hours'].median())

# Fill 0 with mode for Financial Stress (there are 4 values so we use mode to keep another value dominance)
test_df['Financial Stress'] = test_df['Financial Stress'].replace(0, test_df['Financial Stress'].mode()[0])

# Fill 0 with mode for Job/Study Satisfaction
test_df['Job/Study Satisfaction'] = test_df['Job/Study Satisfaction'].replace(0, test_df['Job/Study Satisfaction'].mode()[0])

# Fill 0 with mode for Work/Academic Pressure
test_df['Work/Academic Pressure'] = test_df['Work/Academic Pressure'].replace(0, test_df['Work/Academic Pressure'].mode()[0])

In [40]:
for col in train_df.columns:
    print(f"Value counts for {col}:")
    print(train_df[col].value_counts())
    print("\n")

Value counts for id:
id
0         1
93804     1
93798     1
93799     1
93800     1
         ..
46894     1
46893     1
46892     1
46891     1
140699    1
Name: count, Length: 140673, dtype: int64


Value counts for Gender:
Gender
Male      77448
Female    63225
Name: count, dtype: int64


Value counts for Age:
Age
56.0    5245
49.0    5098
38.0    4564
53.0    4526
57.0    4395
47.0    4199
46.0    4079
54.0    3928
51.0    3926
18.0    3920
43.0    3855
59.0    3781
48.0    3694
45.0    3535
20.0    3515
50.0    3439
42.0    3393
24.0    3352
28.0    3338
39.0    3261
44.0    3217
41.0    3140
33.0    3140
29.0    3096
40.0    3094
25.0    2930
58.0    2930
23.0    2898
55.0    2846
35.0    2834
21.0    2741
37.0    2722
36.0    2703
19.0    2634
27.0    2612
52.0    2589
34.0    2576
31.0    2559
60.0    2501
26.0    2107
22.0    2066
30.0    1911
32.0    1784
Name: count, dtype: int64


Value counts for Working Professional or Student:
Working Professional or Student
Working Profe

## Encoding

In [41]:
label_encoder = LabelEncoder()
ordinal_encoder = OrdinalEncoder()

In [42]:
binary_columns = ['Gender', 'Working Professional or Student', 'Have you ever had suicidal thoughts ?', 'Family History of Mental Illness']
ordinal_columns = ['Dietary Habits', 'Financial Stress']

In [43]:
for col in binary_columns:
    train_df[col] = label_encoder.fit_transform(train_df[col])

In [44]:
for col in ordinal_columns:
    train_df[col] = ordinal_encoder.fit_transform(train_df[[col]])

In [45]:
for col in binary_columns:
    test_df[col] = label_encoder.fit_transform(test_df[col])

In [46]:
for col in ordinal_columns:
    test_df[col] = ordinal_encoder.fit_transform(test_df[[col]])

In [47]:
train_df.head()

id  Gender   Age  Working Professional or Student  Sleep Duration  \
0   0       0  49.0                                1             9.0   
1   1       1  26.0                                1             4.0   
2   2       1  33.0                                0             5.5   
3   3       1  22.0                                1             4.0   
4   4       0  30.0                                1             5.5   

   Dietary Habits  Have you ever had suicidal thoughts ?  Work/Study Hours  \
0             0.0                                      0               1.0   
1             2.0                                      1               7.0   
2             0.0                                      1               3.0   
3             1.0                                      1              10.0   
4             2.0                                      1               9.0   

   Financial Stress  Family History of Mental Illness  Depression  \
0               1.0                                 0           0   
1               2.0                                 0           1   
2               0.0                                 0           1   
3               0.0                                 1           1   
4               3.0                                 1           0   

   Job/Study Satisfaction  Work/Academic Pressure  
0                     2.0                     5.0  
1                     3.0                     4.0  
2                     2.0                     5.0  
3                     1.0                     5.0  
4                     1.0                     1.0

In [48]:
test_df.head()

id  Gender   Age  Working Professional or Student  Sleep Duration  \
0  140700       1  53.0                                1             4.0   
1  140701       0  58.0                                1             4.0   
2  140702       1  53.0                                1             7.5   
3  140703       0  23.0                                0             9.0   
4  140704       1  47.0                                1             7.5   

   Dietary Habits  Have you ever had suicidal thoughts ?  Work/Study Hours  \
0             1.0                                      0               9.0   
1             1.0                                      0               6.0   
2             1.0                                      1              12.0   
3             1.0                                      1              10.0   
4             1.0                                      1               3.0   

   Financial Stress  Family History of Mental Illness  Job/Study Satisfaction  \
0               2.0                                 1                     5.0   
1               3.0                                 0                     4.0   
2               3.0                                 0                     1.0   
3               3.0                                 0                     1.0   
4               3.0                                 0                     5.0   

   Work/Academic Pressure  
0                     2.0  
1                     2.0  
2                     4.0  
3                     5.0  
4                     5.0

# Save Clean Dataset

In [49]:
train_df.to_csv('../data/train_clean.csv', index=False)
test_df.to_csv('../data/test_clean.csv', index=False)